Market decentralized resolver

tworec opened the pull request:
reslover resolves #295 ;)

It glues together changes from #338, #346, #375 and #399.

It also:
- fix query events bugs
- reshapes tests
- introduces DiscoveryBuilder which might be generic solution maybe for Negotiators

nieznanysprawiciel commented on core/market/decentralized/src/matcher/resolver.rs:
```diff
@@ -1 +1,122 @@
+        })
+    }
+
+    pub fn incoming_offer(&self, subscription_id: &str) {
```
Consider using SubscriptionId type in all our not public interfaces.

nieznanysprawiciel commented on core/market/decentralized/tests/test_initial_proposal.rs:
```diff
@@ -155,62 +203,43 @@ mod tests {
-            _ => panic!("Negative maxEvents - expected error"),
-        };
 
-        // Negative timeout should be treated as immediate checking events and return.
-        let events = tokio::time::timeout(
-            Duration::from_millis(20),
-            market1
-                .requestor_engine
-                .query_events(&demand_id, -5.0, None),
-        )
-        .await??;
```
Why did you removed this case?

tworec replied:
I've removed because we need to wait a bit for resolver to process subscriptions. Haven't considered to obviously wait before calling it :) restoring

nieznanysprawiciel commented on core/market/decentralized/tests/test_resolver.rs:
```diff
@@ -0,0 +1,151 @@
+    async fn test_single_not_resolve_offer() -> Result<(), anyhow::Error> {
+        // given
+        let _ = env_logger::builder().try_init();
+        let mut network = MarketsNetwork::new("test_single_resolve_offer")
```
`test_single_resolve_offer` differs from function name `test_single_not_resolve_offer`. For purpose?

tworec replied:
thx

nieznanysprawiciel commented on core/market/decentralized/tests/test_resolver.rs:
```diff
@@ -0,0 +1,151 @@
+        let demand = requestor.subscribe_demand(&sample_demand(), &id3).await?;
+
+        // then: It should be resolved on Requestor two times
+        let listener = network.get_event_listeners("Requestor-1");
+        let proposal = timeout200ms(listener.proposal_rx.recv()).await?.unwrap();
+        assert_eq!(proposal.offer, offer1);
+        assert_eq!(proposal.demand, demand);
+        let proposal = timeout200ms(listener.proposal_rx.recv()).await?.unwrap();
+        assert_eq!(proposal.offer, offer2);
+        assert_eq!(proposal.demand, demand);
```
Is ordering deterministic? Or at least - should test depend on this ordering?

nieznanysprawiciel replied:
But we can leave it

tworec replied:
again good catch. It caused an error and I've fixed it

nieznanysprawiciel commented on core/market/decentralized/tests/utils/mock_net.rs:
```diff
@@ -163,6 +85,78 @@ impl MockNet {
+            let msg = serde_json::to_vec(&msg_json).unwrap();
+            let topic = msg_json.topic().to_owned();
+            let endpoints = bcast.resolve(&caller, &topic);
+
+            log::debug!("BCasting on {} to {:?} from {}", topic, endpoints, caller);
+            for endpoint in endpoints {
+                let addr = format!("{}/{}", endpoint, bcast_service_id);
+                log::debug!("BCasting on {} to {}", topic, addr);
+                let caller = caller.clone();
+                let msg = msg.clone();
+                Arbiter::spawn(async move {
+                    let _ = local_bus::send(addr.as_ref(), &caller, msg.as_ref()).await;
+                });
+            }
+            async move { Ok(Vec::from(resp.as_ref())) }
+        });
+
+        local_bus::subscribe(FROM_BUS_ID, move |caller: &str, addr: &str, msg: &[u8]| {
+            let mock_net = MockNet::default();
+            let data = Vec::from(msg);
+            let caller = caller.to_string();
+            let addr = addr.to_string();
+
+            async move {
+                let local_addr = mock_net
+                    .translate_address(addr)
+                    .await
+                    .map_err(|e| Error::GsbBadRequest(e.to_string()))?;
+
+                log::debug!(
+                    "[MockNet] Sending message from [{}], to address [{}].",
+                    &caller,
+                    &local_addr
+                );
+                Ok(local_bus::send(&local_addr, &caller, &data).await?)
+            }
+        });
+    }
+}
+
```
Is it just copy, or should I read it?

tworec replied:
copy

tworec commented on core/market/decentralized/src/db/dao/events.rs:
```diff
@@ -12,16 +12,11 @@ use crate::db::DbResult;
     #[error("Subscription [{0}] expired.")]
     SubscriptionExpired(SubscriptionId),
-    #[error(transparent)]
+    #[error("Failed to get events from database. Error: {0}.")]
```
so do I, that's why I changed your code ;)

nieznanysprawiciel replied:
:facepalm:

tworec commented on core/market/decentralized/src/db/dao/events.rs:
```diff
@@ -70,31 +65,17 @@ impl<'c> EventsDao<'c> {
+            if !events.is_empty() {
+                let ids = events.iter().map(|event| event.id).collect::<Vec<_>>();
+                if !ids.is_empty() {
+                    diesel::delete(dsl::market_event.filter(dsl::id.eq_any(ids))).execute(conn)?;
```
Diesel statements map quite literally to SQL

tworec commented on core/market/decentralized/src/db/dao/events.rs:
```diff
@@ -70,31 +65,17 @@ impl<'c> EventsDao<'c> {
+            // Remove returned events from queue.
+            if !events.is_empty() {
+                let ids = events.iter().map(|event| event.id).collect::<Vec<_>>();
+                if !ids.is_empty() {
```
will remove this check

tworec commented on core/market/decentralized/src/lib.rs:
```diff
@@ -23,14 +25,3 @@ pub mod migrations {
     #[derive(diesel_migrations::EmbedMigrations)]
     struct _Dummy;
 }
-
-/// These exports are expected to be used only in tests.
-pub mod testing {
-    pub use super::db::dao::{DemandDao, OfferDao};
-    pub use super::db::models::{Demand, Offer, SubscriptionId, SubscriptionParseError};
-    pub use super::matcher::{DemandError, MatcherError, OfferError};
-    pub use super::matcher::{DraftProposal, SubscriptionStore};
-    pub use super::negotiation::QueryEventsError;
-    pub use super::negotiation::{ProviderBroker, RequestorBroker};
-    pub use super::protocol::negotiation;
-}
```
because I've moved `mock_offer.rs` from `tests` to be testing sub-module and wanted to hide it in subdir.

tworec commented on core/market/decentralized/src/market.rs:
```diff
@@ -55,7 +56,7 @@ impl MarketService {
         let (matcher, listeners) = Matcher::new(db)?;
         let provider_engine = ProviderBroker::new(db.clone())?;
-        let requestor_engine = RequestorBroker::new(db.clone(), listeners.proposal_receiver)?;
+        let requestor_engine = RequestorBroker::new(db.clone(), listeners.proposal_rx)?;
```
`tx, rx` is a de-facto naming standard for the `sender, receiver`. See
```
grep --include '*.rs' -rn 'let[^=]*rx\>' agent core exe-unit service-bus test-utils utils
```
which gives 34 results across all yagna repo, comparing to only 6 from
```
grep --include '*.rs' -rn 'let[^=]*receiver\>' agent core exe-unit service-bus test-utils utils
```
3 of which are from the decentralized market still

nieznanysprawiciel replied:
I know and it's terrible. At least don't change my names

tworec replied:
it is also in code examples within Tokio eg.
https://docs.rs/tokio/0.2.21/tokio/sync/mpsc/struct.UnboundedReceiver.html

I'm reverting within your code if it is hurting you so much

tworec commented on core/market/decentralized/src/matcher.rs:
```diff
@@ -148,26 +146,27 @@ impl Matcher {
         .store_offer(msg.offer)
         .await
         .map(|propagate| match propagate {
-            true => Propagate::Yes,
+            true => {
+                resolver.receive(subscription).unwrap();
+                Propagate::Yes
+            }
```
will fix

tworec replied:
This flaw was introduced originally in #290 :-P
https://github.com/golemfactory/yagna/pull/290/files#diff-c373b03f716ab716b09699faa1c50599R227-R255

tworec commented on core/market/decentralized/src/matcher/error.rs:
```diff
@@ -56,6 +64,16 @@ pub enum MatcherInitError {
 
+#[derive(thiserror::Error, Debug)]
+pub enum ResolverError {
+    #[error("Failed resolve matching relation for {0:?} and {1:?}.")]
+    MatchingError(Offer, Demand),
```
yes, this uses debug `:?`

nieznanysprawiciel replied:
I don't think, it's good idea. I wouldn't print such an error later. It pollutes whole log and we can't see details whatsoever

nieznanysprawiciel replied:
Hmm if you want to have them printed somewhere, you could make log::trace in place where this error happens, but return in error only ids

tworec replied:
ok

tworec commented on core/market/decentralized/src/matcher/error.rs:
```diff
@@ -56,6 +64,16 @@ pub enum MatcherInitError {
+    #[error("Failed resolve matching relation for {0:?} and {1:?}.")]
+    MatchingError(Offer, Demand),
+    #[error("Failed to process incoming subscription {0:?}")]
+    SendError(#[from] tokio::sync::mpsc::error::SendError<Subscription>),
```
Good point. There will be no subscription id which caused the problem. I Will enrich this err msg it.

tworec replied:
removed this enum option altogether

tworec commented on core/market/decentralized/src/matcher/resolver.rs:
```diff
@@ -1 +1,136 @@
+        mut subscription_rx: UnboundedReceiver<Subscription>,
+    ) {
+        while let Some(subscription) = subscription_rx.recv().await {
+            log::debug!("processing incoming subscription {:?}", subscription);
```
this was on purpose. Subscription does not support display. And it is an enum with name of subscription id: Offer or Demand

tworec commented on core/market/decentralized/src/matcher/resolver.rs:
```diff
@@ -1 +1,136 @@
+            log::debug!("processing incoming subscription {:?}", subscription);
+            match subscription {
+                Subscription::Offer(id) => {
+                    log::info!("TODO: resolve new Offer: {:?}", id);
```
yup, it was leftover.

tworec commented on core/market/decentralized/src/matcher/store.rs:
```diff
@@ -58,6 +58,37 @@ impl SubscriptionStore {
+            .collect())
+    }
+
+    pub async fn get_offers_before(&self, demand: &Demand) -> Result<Vec<Offer>, OfferError> {
+        let now = Utc::now().naive_utc();
+        Ok(self
+            .db
+            .as_dao::<OfferDao>()
+            .get_offers_before(demand.insertion_ts.unwrap(), now)
+            .await
+            .map_err(|e| OfferError::GetMany(e))?)
+    }
+
```
Good point.

The intent was to show the purpose or the context in which this is usable. `insertion_ts` is designed for that single purpose to be compared against other Demand[s], not between Offers themselves. 

But now I see I can go even further and enclose getting single Demand and all Offers before it within single offer store fn. Since it is the only usecase for it and will illustrate the intent clearly. Furthermore, it will ensure that `unwrap` is safe.

tworec replied:
but for the time being, I'm simplifying the API as you've suggested

tworec commented on core/market/decentralized/src/matcher/store.rs:
```diff
@@ -58,6 +58,37 @@ impl SubscriptionStore {
+        Ok(self
+            .db
+            .as_dao::<OfferDao>()
+            .get_offers(match id {
+                Some(ident) => Some(ident.identity),
+                _ => None,
+            })
```
nice :) thx

tworec replied:
This is @jiivan code. As I've merged to his PR, but still I will apply this simplification.

tworec commented on core/market/decentralized/src/matcher/resolver.rs:
```diff
@@ -1 +1,136 @@
+            )
+        };
+    }
+}
```
Thanks for this comment. This enabled me to find a bug, bc current code will break the whole matching when one pair (offer, demand) matching causes an error, and this should not be broken but to be continued for all remaining pairs.

tworec commented on core/market/decentralized/src/negotiation/errors.rs:
```diff
@@ -31,13 +29,3 @@ pub enum QueryEventsError {
 
 #[derive(Error, Debug)]
 pub enum ProposalError {}
-
-impl From<TakeEventsError> for QueryEventsError {
-    fn from(e: TakeEventsError) -> Self {
-        match e {
-            TakeEventsError::SubscriptionExpired(id) => QueryEventsError::SubscriptionExpired(id),
-            TakeEventsError::SubscriptionNotFound(id) => QueryEventsError::Unsubscribed(id),
-            _ => QueryEventsError::FailedGetEvents(e),
-        }
-    }
-}
```
I've changed it because it has flaw. If we've got `TakeEventsError::SubscriptionNotFound` we cannot translate it into `QueryEventsError::Unsubscribed` because it is not only possibility. Subscription might never existed. This change is not only for code brevity, but also for correctness

nieznanysprawiciel replied:
Than you should rather change QueryEventsError to `SubscriptionNotFound`, because we can never return `QueryEventsError::Unsubscribed`

nieznanysprawiciel replied:
I don't like this pattern of embedding errors in errors in every place possible. It makes sense in some places, but in some doesn't. QueryEventsError should give very specific errors without need to dispatch nesting.

tworec replied:
> Than you should rather change QueryEventsError to `SubscriptionNotFound`, because we can never return `QueryEventsError::Unsubscribed`
yes, we can and it is illustrated two times in test


> I don't like this pattern of embedding errors in errors in every place possible. It makes sense in some places, but in some doesn't. QueryEventsError should give very specific errors without need to dispatch nesting.

Here it seems to me it is one of the case when embedding works very well, because otherwise you need to repeat all patterns from `TakeEventsError` in `QueryEventsError`, isn't it?

nieznanysprawiciel replied:
Repeating isn't a problem for me. Ugly api is a problem.

And what are patterns on which caller wants to match against here?
_NotFound_, _Unsubscribed_, _Expired_ are fundamental errors and caller can react somehow to them.
Other errors aren't so important, because caller could threat them as "some internal error" and he can't do anything more with this. He can just print message and wait for human to read it ;)

In this code the most important errors are hidden behind strange enum `TakeEventsError`. `TakeEventsError` tells him nothing, because he already knows that querying events failed. So he must dive deeper, to know what happened.

Embedding makes sense, when we deal with errors from internal layers. We want to wrap them with higher level errors, to add more meaningful message. When there are multiple layers, we have full path of errors that caused problem.
In such case wrapping adds information. Here we take information.

nieznanysprawiciel replied:
I think we could leave it, but maybe we should make some meeting to establish common error handling practices, that would satisfy us all

tworec commented on core/market/decentralized/src/protocol/discovery.rs:
```diff
@@ -242,7 +231,7 @@ pub enum Propagate {
 }
 
-#[derive(Clone, Serialize, Deserialize)]
+#[derive(Clone, Serialize, Deserialize, Debug)]
```
probably some debugging. It is a leftover. Removing

tworec commented on core/market/decentralized/src/protocol/discovery/builder.rs:
```diff
@@ -0,0 +1,172 @@
+            .get(&TypeId::of::<T>())
+            .expect("data handler needs data");
+
+        // let data = *(boxed as &Box<dyn Any + 'static>).downcast::<&T>().unwrap();
```
nope, removing

tworec commented on core/market/decentralized/src/protocol/discovery/builder.rs:
```diff
@@ -0,0 +1,172 @@
+        let boxed = self
+            .data
+            .get(&TypeId::of::<T>())
+            .expect("data handler needs data");
```
Because this only can happen when the Market developer (one of us) is misusing the builder (forgotten to specify data). It is not user-facing error. it should fail fast during initial testing or demo

nieznanysprawiciel replied:
And the same happens if you return error from initialization

tworec replied:
Nope, in intialisation you are also handling GSB errors... it might be user-facing

nieznanysprawiciel replied:
Ok, if you insists ;)

tworec commented on core/market/decentralized/src/rest_api/error.rs:
```diff
@@ -69,8 +79,11 @@ impl ResponseError for QueryEventsError {
         match self {
-            QueryEventsError::Unsubscribed(_) => HttpResponse::NotFound().json(msg),
-            QueryEventsError::SubscriptionExpired(_) => HttpResponse::NotFound().json(msg),
+            QueryEventsError::Unsubscribed(_)
+            | QueryEventsError::TakeEventsError(TakeEventsError::SubscriptionNotFound(_))
+            | QueryEventsError::TakeEventsError(TakeEventsError::SubscriptionExpired(_)) => {
+                HttpResponse::NotFound().json(msg)
+            }
```
I've chosen not to redefine  those two `TakeEventsError` values within  `QueryEventsError`... for the code brevity

tworec commented on core/market/decentralized/src/rest_api/error.rs:
```diff
@@ -40,6 +42,14 @@ impl ResponseError for MatcherError {
 
 impl ResponseError for NegotiationError {}
 
+impl ResponseError for ResolverError {
+    fn error_response(&self) -> HttpResponse {
+        match self {
+            _ => HttpResponse::InternalServerError().json(ErrorMessage::new(self.to_string())),
+        }
+    }
+}
+
```
Good catch! Thx

tworec commented on core/market/decentralized/tests/test_initial_proposal.rs:
```diff
@@ -2,21 +2,40 @@ mod utils;
+    /// No events available for not existent subscription.
+    #[cfg_attr(not(feature = "market-test-suite"), ignore)]
+    #[actix_rt::test]
+    async fn test_query_events_non_existent_subscription() -> Result<(), anyhow::Error> {
```
I like naming and isolation of edge cases

tworec commented:
@nieznanysprawiciel pls review again

nieznanysprawiciel commented on core/market/decentralized/src/db/dao/offer.rs:
```diff
@@ -58,6 +92,25 @@ impl<'c> OfferDao<'c> {
         .await
     }
 
+    pub async fn checked_insert(
```
All inserts should be checked. No need to highlight this in name

tworec replied:
initial should not I believe...

nieznanysprawiciel replied:
Probably we should have only one function. Caller decides if this failing check is an error or valid situation.
But doesn't matter. Let's merge this quickly

tworec replied:
Ok, now there is a single insert function which returns resulting Offer with `insertion_ts`.

nieznanysprawiciel commented on core/market/decentralized/src/matcher.rs:
```diff
@@ -148,26 +145,27 @@ impl Matcher {
+        .store
+        .save_offer(msg.offer)
         .await
         .map(|propagate| match propagate {
```
This should not be named `propagate`, but probably `inserted` or something like this. Save_offer doesn't return info, if we should propagate anything

tworec replied:
true

nieznanysprawiciel commented on core/market/decentralized/src/matcher/error.rs:
```diff
@@ -56,6 +62,14 @@ pub enum MatcherInitError {
+#[derive(thiserror::Error, Debug)]
+pub enum ResolverError {
+    #[error(transparent)]
+    OfferError(#[from] OfferError),
```
OfferError has ResolverError and ResolverError has OfferError

tworec replied:
This shouldn't be possible. `MatcherError` has `ResolveError` which has `OfferError`

tworec replied:
I've tested and such setup results in compile error:
```
error[E0391]: cycle detected when processing `matcher::error::OfferError`
```

tworec commented:
A few improvements:
- cleaned unused errors
- divided huge `OfferError` into specialized errors eq. `QueryOfferError`
- moved unsubscribe offer logic to `SubscriptionStore` which is capable of deciding when to remove Offer and when to just mark it as unsubscribed.

@nieznanysprawiciel pls re-review
